File: packages/executable-stories-junit5/src/main/kotlin/dev/executablestories/junit5/Story.kt
package dev.executablestories.junit5

import java.util.function.Supplier

/**
 * Static fluent API for defining BDD stories within JUnit 5 tests.
 * Uses ThreadLocal to maintain per-test story state.
 */
class Story private constructor() {
    companion object {
        private val CONTEXT = ThreadLocal<StoryContext>()

        /**
         * Declarations keyed by the class that made them, so a re-declaration
         * reads the way it does in source order.
         */
        private val FEATURES = java.util.concurrent.ConcurrentHashMap<String, Map<String, Any?>>()

        // ====================================================================
        // Init
        // ====================================================================

        @JvmStatic
        fun init(scenario: String) {
            val ctx = StoryContext(scenario)
            CONTEXT.set(ctx)
            bridgeOtel(ctx)
        }

        @JvmStatic
        fun init(
            scenario: String,
            vararg tags: String,
        ) {
            val ctx = StoryContext(scenario, *tags)
            CONTEXT.set(ctx)
            bridgeOtel(ctx)
        }

        /**
         * Declares a scenario that is specified but not built yet. It appears in
         * the report marked "planned" and stops being planned once someone writes
         * it as a real story with [init].
         *
         * ```kotlin
         * @Test
         * fun `checkout is blocked for a suspended account`() {
         *     Story.planned("checkout is blocked for a suspended account")
         * }
         * ```
         *
         * `@Disabled` means "do not run this now", which is a different claim from
         * "we have not built this yet", so this does not disable the test for you.
         */
        @JvmStatic
        fun planned(
            scenario: String,
            vararg tags: String,
        ) {
            val ctx = StoryContext(scenario, *tags)
            ctx.planned = true
            CONTEXT.set(ctx)
        }

        @JvmStatic
        fun withTraceUrlTemplate(template: String) {
            requireContext().traceUrlTemplate = template
        }

        // ====================================================================
        // Step methods
        // ====================================================================

        @JvmStatic
        fun given(text: String) {
            requireContext().addStep("Given", text)
        }

        @JvmStatic
        fun given(
            text: String,
            vararg docs: DocEntry,
        ) {
            requireContext().addStep("Given", text, *docs)
        }

        @JvmStatic
        @JvmName("when")
        fun `when`(text: String) {
            requireContext().addStep("When", text)
        }

        @JvmStatic
        @JvmName("when")
        fun `when`(
            text: String,
            vararg docs: DocEntry,
        ) {
            requireContext().addStep("When", text, *docs)
        }

        @JvmStatic
        fun then(text: String) {
            requireContext().addStep("Then", text)
        }

        @JvmStatic
        fun then(
            text: String,
            vararg docs: DocEntry,
        ) {
            requireContext().addStep("Then", text, *docs)
        }

        @JvmStatic
        fun and(text: String) {
            requireContext().addStep("And", text)
        }

        @JvmStatic
        fun and(
            text: String,
            vararg docs: DocEntry,
        ) {
            requireContext().addStep("And", text, *docs)
        }

        @JvmStatic
        fun but(text: String) {
            requireContext().addStep("But", text)
        }

        @JvmStatic
        fun but(
            text: String,
            vararg docs: DocEntry,
        ) {
            requireContext().addStep("But", text, *docs)
        }

        // ====================================================================
        // AAA Pattern Aliases
        // ====================================================================

        @JvmStatic
        fun arrange(text: String) {
            requireContext().addStep("Given", text)
        }

        @JvmStatic
        fun arrange(
            text: String,
            vararg docs: DocEntry,
        ) {
            requireContext().addStep("Given", text, *docs)
        }

        @JvmStatic
        fun act(text: String) {
            requireContext().addStep("When", text)
        }

        @JvmStatic
        fun act(
            text: String,
            vararg docs: DocEntry,
        ) {
            requireContext().addStep("When", text, *docs)
        }

        @JvmStatic
        fun assertThat(text: String) {
            requireContext().addStep("Then", text)
        }

        @JvmStatic
        fun assertThat(
            text: String,
            vararg docs: DocEntry,
        ) {
            requireContext().addStep("Then", text, *docs)
        }

        // ====================================================================
        // Additional Aliases
        // ====================================================================

        @JvmStatic
        fun setup(text: String) {
            requireContext().addStep("Given", text)
        }

        @JvmStatic
        fun setup(
            text: String,
            vararg docs: DocEntry,
        ) {
            requireContext().addStep("Given", text, *docs)
        }

        @JvmStatic
        fun context(text: String) {
            requireContext().addStep("Given", text)
        }

        @JvmStatic
        fun context(
            text: String,
            vararg docs: DocEntry,
        ) {
            requireContext().addStep("Given", text, *docs)
        }

        @JvmStatic
        fun execute(text: String) {
            requireContext().addStep("When", text)
        }

        @JvmStatic
        fun execute(
            text: String,
            vararg docs: DocEntry,
        ) {
            requireContext().addStep("When", text, *docs)
        }

        @JvmStatic
        fun action(text: String) {
            requireContext().addStep("When", text)
        }

        @JvmStatic
        fun action(
            text: String,
            vararg docs: DocEntry,
        ) {
            requireContext().addStep("When", text, *docs)
        }

        @JvmStatic
        fun verify(text: String) {
            requireContext().addStep("Then", text)
        }

        @JvmStatic
        fun verify(
            text: String,
            vararg docs: DocEntry,
        ) {
            requireContext().addStep("Then", text, *docs)
        }

        // ====================================================================
        // Doc methods
        // ====================================================================

        @JvmStatic
        fun note(text: String): DocEntry {
            val entry = DocEntry.note(text)
            requireContext().addDoc(entry)
            return entry
        }

        @JvmStatic
        fun tag(vararg names: String): DocEntry {
            val entry = DocEntry.tag(*names)
            requireContext().addDoc(entry)
            return entry
        }

        @JvmStatic
        fun kv(
            label: String,
            value: Any?,
        ): DocEntry {
            val entry = DocEntry.kv(label, value)
            requireContext().addDoc(entry)
            return entry
        }

        @JvmStatic
        fun json(
            label: String,
            value: Any?,
        ): DocEntry {
            val entry = DocEntry.json(label, value)
            requireContext().addDoc(entry)
            return entry
        }

        /** Snapshot of the world at the current step (kind=state). */
        @JvmStatic
        @JvmOverloads
        fun state(
            value: Any?,
            label: String? = null,
        ): DocEntry {
            val entry = DocEntry.state(value, label)
            requireContext().addDoc(entry)
            return entry
        }

        @JvmStatic
        fun code(
            label: String,
            content: String,
        ): DocEntry {
            val entry = DocEntry.code(label, content, null)
            requireContext().addDoc(entry)
            return entry
        }

        @JvmStatic
        fun code(
            label: String,
            content: String,
            lang: String,
        ): DocEntry {
            val entry = DocEntry.code(label, content, lang)
            requireContext().addDoc(entry)
            return entry
        }

        @JvmStatic
        fun table(
            label: String,
            columns: Array<String>,
            rows: Array<Array<String>>,
        ): DocEntry {
            val entry = DocEntry.table(label, columns, rows)
            requireContext().addDoc(entry)
            return entry
        }

        @JvmStatic
        fun link(
            label: String,
            url: String,
        ): DocEntry {
            val entry = DocEntry.link(label, url)
            requireContext().addDoc(entry)
            return entry
        }

        @JvmStatic
        fun section(
            title: String,
            markdown: String,
        ): DocEntry {
            val entry = DocEntry.section(title, markdown)
            requireContext().addDoc(entry)
            return entry
        }

        @JvmStatic
        fun mermaid(code: String): DocEntry {
            val entry = DocEntry.mermaid(code, null)
            requireContext().addDoc(entry)
            return entry
        }

        @JvmStatic
        fun mermaid(
            code: String,
            title: String,
        ): DocEntry {
            val entry = DocEntry.mermaid(code, title)
            requireContext().addDoc(entry)
            return entry
        }

        @JvmStatic
        fun screenshot(path: String): DocEntry {
            val entry = DocEntry.screenshot(path, null)
            requireContext().addDoc(entry)
            return entry
        }

        @JvmStatic
        fun screenshot(
            path: String,
            alt: String,
        ): DocEntry {
            val entry = DocEntry.screenshot(path, alt)
            requireContext().addDoc(entry)
            return entry
        }

        /**
         * Attach an embedded-HTML doc entry. Exactly one of [path], [url], or
         * [content] must be non-null.
         */
        @JvmStatic
        @JvmOverloads
        fun html(
            path: String? = null,
            url: String? = null,
            content: String? = null,
            title: String? = null,
            height: Any? = null,
        ): DocEntry {
            val entry = DocEntry.html(path, url, content, title, height)
            requireContext().addDoc(entry)
            return entry
        }

        @JvmStatic
        fun custom(
            type: String,
            data: Any?,
        ): DocEntry {
            val entry = DocEntry.custom(type, data)
            requireContext().addDoc(entry)
            return entry
        }

        // ====================================================================
        // Ticket methods
        // ====================================================================

        @JvmStatic
        fun ticket(id: String) {
            requireContext().tickets.add(Ticket(id))
        }

        @JvmStatic
        fun ticket(
            id: String,
            url: String,
        ) {
            requireContext().tickets.add(Ticket(id, url))
        }

        @JvmStatic
        fun ticket(ticket: Ticket) {
            requireContext().tickets.add(ticket)
        }

        /** Declare the product-code paths/globs this story exercises. */
        @JvmStatic
        fun covers(vararg paths: String) {
            requireContext().covers.addAll(paths.toList())
        }

        // ====================================================================
        // Attachment methods
        // ====================================================================

        @JvmStatic
        fun attach(
            name: String,
            mediaType: String,
            path: String,
        ) {
            requireContext().addAttachment(name, mediaType, path, null, null, null, null)
        }

        @JvmStatic
        fun attachInline(
            name: String,
            mediaType: String,
            body: String,
            encoding: String,
        ) {
            requireContext().addAttachment(name, mediaType, null, body, encoding, null, null)
        }

        // ====================================================================
        // OTel span attachment
        // ====================================================================

        @JvmStatic
        fun attachSpans(spans: List<Any>) {
            requireContext().otelSpans = spans
        }

        // ====================================================================
        // Step timing
        // ====================================================================

        @JvmStatic
        fun startTimer(): Int = requireContext().startTimer()

        @JvmStatic
        fun endTimer(token: Int) {
            requireContext().endTimer(token)
        }

        // ====================================================================
        // Wrapped step execution
        // ====================================================================

        @JvmStatic
        fun fn(
            keyword: String,
            text: String,
            body: Runnable,
        ) {
            val ctx = requireContext()
            ctx.addStep(keyword, text)
            val step = ctx.currentStep!!
            step.wrapped = true

            val start = System.nanoTime()
            try {
                body.run()
            } finally {
                step.durationMs = (System.nanoTime() - start) / 1_000_000.0
            }
        }

        @JvmStatic
        fun <T> fn(
            keyword: String,
            text: String,
            body: Supplier<T>,
        ): T {
            val ctx = requireContext()
            ctx.addStep(keyword, text)
            val step = ctx.currentStep!!
            step.wrapped = true

            val start = System.nanoTime()
            try {
                return body.get()
            } finally {
                step.durationMs = (System.nanoTime() - start) / 1_000_000.0
            }
        }

        @JvmStatic
        fun expect(
            text: String,
            body: Runnable,
        ) {
            fn("Then", text, body)
        }

        @JvmStatic
        fun <T> expect(
            text: String,
            body: Supplier<T>,
        ): T = fn("Then", text, body)

        // ====================================================================
        // OTel Bridge
        // ====================================================================

        private fun bridgeOtel(ctx: StoryContext) {
            try {
                val spanClass = Class.forName("io.opentelemetry.api.trace.Span")
                val currentMethod = spanClass.getMethod("current")
                val span = currentMethod.invoke(null) ?: return

                val spanContextMethod = spanClass.getMethod("getSpanContext")
                val spanContext = spanContextMethod.invoke(span) ?: return

                val spanContextClass = spanContext.javaClass
                val getTraceId = spanContextClass.getMethod("getTraceId")
                val getSpanId = spanContextClass.getMethod("getSpanId")
                val isValid = spanContextClass.getMethod("isValid")

                if (isValid.invoke(spanContext) != true) return

                val traceId = getTraceId.invoke(spanContext) as? String ?: return
                val spanId = getSpanId.invoke(spanContext) as? String ?: return

                if (traceId == "00000000000000000000000000000000") return

                // OTel -> Story: capture traceId in structured meta
                ctx.meta["otel"] = mapOf("traceId" to traceId, "spanId" to spanId)

                // OTel -> Story: inject human-readable doc entries
                ctx.docs.add(DocEntry.kv("Trace ID", traceId))

                val template = ctx.traceUrlTemplate ?: System.getenv("OTEL_TRACE_URL_TEMPLATE")
                if (!template.isNullOrEmpty()) {
                    val url = template.replace("{traceId}", traceId)
                    ctx.docs.add(DocEntry.link("View Trace", url))
                }

                // Story -> OTel: enrich active span with story attributes
                val setAttributeStr = spanClass.getMethod("setAttribute", String::class.java, String::class.java)
                setAttributeStr.invoke(span, "story.scenario", ctx.scenario)

                if (ctx.tags.isNotEmpty()) {
                    setOtelArrayAttribute(spanClass, span, "story.tags", ctx.tags.toList())
                }

                if (ctx.tickets.isNotEmpty()) {
                    setOtelArrayAttribute(spanClass, span, "story.tickets", ctx.tickets.map { it.id })
                }
            } catch (_: ClassNotFoundException) {
                // OTel API not on classpath - no-op
            } catch (_: Exception) {
                // OTel not available - no-op
            }
        }

        private fun setOtelArrayAttribute(
            spanClass: Class<*>,
            span: Any,
            key: String,
            values: List<String>,
        ) {
            try {
                val attributeKeyClass = Class.forName("io.opentelemetry.api.common.AttributeKey")
                val stringArrayKeyMethod = attributeKeyClass.getMethod("stringArrayKey", String::class.java)
                val attrKey = stringArrayKeyMethod.invoke(null, key)
                val setAttributeKey = spanClass.getMethod("setAttribute", attributeKeyClass, Any::class.java)
                setAttributeKey.invoke(span, attrKey, values)
            } catch (_: Exception) {
                // array attributes not available
            }
        }

        // ====================================================================
        // Internal
        // ====================================================================

        /**
         * Declare what a class's scenarios are for, ahead of the examples.
         *
         * Scenarios say what the system does. A declaration says why the
         * feature exists and who it serves, so a reader meets the intent
         * before the examples. Call it once per test class, from `@BeforeAll`:
         *
         * ```kotlin
         * companion object {
         *     @BeforeAll
         *     @JvmStatic
         *     fun declareFeature() {
         *         Story.feature(
         *             title = "Anyone can do arithmetic without a calculator app",
         *             kind = "ability",
         *             narrative = "Switching apps for a quick sum loses your place.",
         *         )
         *     }
         * }
         * ```
         *
         * The JVM gives us no source path, so the declaring class is the key
         * the report groups by, taken from the call stack.
         */
        @JvmStatic
        @JvmOverloads
        fun feature(
            title: String,
            kind: String? = null,
            narrative: String? = null,
            tags: List<String>? = null,
            glossary: Map<String, String>? = null,
        ) {
            val key = callerClassName() ?: return
            val declared = LinkedHashMap<String, Any?>()
            declared["sourceFile"] = key
            declared["title"] = title
            if (kind != null) declared["kind"] = kind
            if (narrative != null) declared["narrative"] = narrative
            if (!tags.isNullOrEmpty()) declared["tags"] = tags
            if (!glossary.isNullOrEmpty()) {
                declared["glossary"] =
                    glossary.map { (term, definition) ->
                        linkedMapOf<String, Any?>("term" to term, "definition" to definition)
                    }
            }
            FEATURES[key] = declared
        }

        @JvmStatic
        internal fun declaredFeatures(): List<Map<String, Any?>> = FEATURES.values.toList()

        /**
         * Class that called into this object, with Kotlin's synthetic suffixes
         * dropped so a declaration made from a companion object still keys to
         * the test class itself.
         */
        private fun callerClassName(): String? {
            val frames = Thread.currentThread().stackTrace
            for (frame in frames) {
                val name = frame.className
                if (name.startsWith("java.") || name.startsWith("dev.executablestories.junit5.Story")) {
                    continue
                }
                return name.substringBefore("\$Companion").substringBefore("\$\$")
            }
            return null
        }

        @JvmStatic
        internal fun getContext(): StoryContext? = CONTEXT.get()

        @JvmStatic
        fun clear() {
            CONTEXT.remove()
        }

        private fun requireContext(): StoryContext =
            CONTEXT.get() ?: throw IllegalStateException(
                "Story.init() must be called before using Story step/doc methods. " +
                    "Did you forget to call Story.init(\"scenario name\") at the start of your test?",
            )
    }
}

File: packages/executable-stories-ruby/lib/executable_stories/story.rb
# frozen_string_literal: true

require "set"

require_relative "types"
require_relative "doc_entry"
require_relative "collector"

module ExecutableStories
  class Story
    attr_accessor :scenario, :steps, :tags, :tickets, :covers, :meta, :docs,
                  :current_step, :seen_primary, :start_time, :end_time,
                  :source_order, :step_counter, :attachments, :active_timers,
                  :timer_counter, :otel_spans, :trace_url_template, :suite_path

    def initialize(scenario, tags: nil, ticket: nil, covers: nil, meta: nil, trace_url_template: nil)
      @scenario = scenario
      @steps = []
      @tags = tags
      @tickets = normalize_tickets(ticket)
      @covers = covers
      @meta = meta&.dup
      @docs = []
      @current_step = nil
      @seen_primary = {}
      @start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC) * 1000.0
      @source_order = Collector.next_order
      @step_counter = 0
      @attachments = []
      @active_timers = {}
      @timer_counter = 0
      @otel_spans = nil
      @trace_url_template = trace_url_template
      @suite_path = nil

      bridge_otel
    end

    def given(text)
      add_step("Given", text)
      self
    end

    def when(text)
      add_step("When", text)
      self
    end

    def then(text)
      add_step("Then", text)
      self
    end

    def and(text)
      add_explicit_step("And", text)
      self
    end

    def but(text)
      add_explicit_step("But", text)
      self
    end

    def arrange(text)
      add_step("Given", text)
      self
    end

    def act(text)
      add_step("When", text)
      self
    end

    def assert_that(text)
      add_step("Then", text)
      self
    end

    def setup(text)
      add_step("Given", text)
      self
    end

    def context(text)
      add_step("Given", text)
      self
    end

    def execute(text)
      add_step("When", text)
      self
    end

    def action(text)
      add_step("When", text)
      self
    end

    def verify(text)
      add_step("Then", text)
      self
    end

    def fn(keyword, text, &body)
      add_step(keyword, text)
      @current_step.wrapped = true

      start_ms = Process.clock_gettime(Process::CLOCK_MONOTONIC) * 1000.0
      begin
        result = body.call
        @current_step.duration_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) * 1000.0) - start_ms
        result
      rescue StandardError => e
        @current_step.duration_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) * 1000.0) - start_ms
        raise e
      end
    end

    def expect(text, &)
      fn("Then", text, &)
    end

    def start_timer
      token = @timer_counter
      @timer_counter += 1

      step_index = @current_step ? @steps.index(@current_step) : nil
      entry = {
        start: Process.clock_gettime(Process::CLOCK_MONOTONIC) * 1000.0,
        step_index: step_index,
        step_id: @current_step&.id,
        consumed: false
      }
      @active_timers[token] = entry
      token
    end

    def end_timer(token)
      entry = @active_timers[token]
      return unless entry && !entry[:consumed]

      entry[:consumed] = true
      duration_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) * 1000.0) - entry[:start]

      step = nil
      step = @steps.find { |s| s.id == entry[:step_id] } if entry[:step_id]
      step = @steps[entry[:step_index]] if !step && entry[:step_index] && entry[:step_index] < @steps.length

      step.duration_ms = duration_ms if step
    end

    def note(text, children: nil)
      entry = DocEntry.note(text, children: children)
      attach_doc(entry)
      entry
    end

    def tag(*names, children: nil)
      entry = DocEntry.tag(*names, children: children)
      attach_doc(entry)
      entry
    end

    def kv(label, value, children: nil)
      entry = DocEntry.kv(label, value, children: children)
      attach_doc(entry)
      entry
    end

    def json(label, value, children: nil)
      entry = DocEntry.json_doc(label, value, children: children)
      attach_doc(entry)
      entry
    end

    def code(label, content, lang: nil, children: nil)
      entry = DocEntry.code(label, content, lang: lang, children: children)
      attach_doc(entry)
      entry
    end

    def table(label, columns, rows, children: nil)
      entry = DocEntry.table(label, columns, rows, children: children)
      attach_doc(entry)
      entry
    end

    def link(label, url, children: nil)
      entry = DocEntry.link(label, url, children: children)
      attach_doc(entry)
      entry
    end

    def section(title, markdown, children: nil)
      entry = DocEntry.section(title, markdown, children: children)
      attach_doc(entry)
      entry
    end

    def mermaid(code, title: nil, children: nil)
      entry = DocEntry.mermaid(code, title: title, children: children)
      attach_doc(entry)
      entry
    end

    def screenshot(path, alt: nil, children: nil)
      entry = DocEntry.screenshot(path, alt: alt, children: children)
      attach_doc(entry)
      entry
    end

    def state(value, label: nil, children: nil)
      entry = DocEntry.state(value, label: label, children: children)
      attach_doc(entry)
      entry
    end

    def html(path: nil, url: nil, content: nil, title: nil, height: nil, children: nil)
      entry = DocEntry.html(path: path, url: url, content: content, title: title, height: height, children: children)
      attach_doc(entry)
      entry
    end

    def custom(type, data, children: nil)
      entry = DocEntry.custom(type, data, children: children)
      attach_doc(entry)
      entry
    end

    def attach(name, media_type, path: nil, body: nil, encoding: nil, charset: nil, file_name: nil)
      att = RawAttachment.new(
        name: name,
        media_type: media_type,
        path: path,
        body: body,
        encoding: encoding,
        charset: charset,
        file_name: file_name
      )
      if @current_step
        idx = @steps.index(@current_step)
        att.step_index = idx if idx
        att.step_id = @current_step.id
      end
      @attachments << att
      self
    end

    def attach_inline(name, media_type, body, encoding: "IDENTITY")
      att = RawAttachment.new(
        name: name,
        media_type: media_type,
        body: body,
        encoding: encoding
      )
      if @current_step
        idx = @steps.index(@current_step)
        att.step_index = idx if idx
        att.step_id = @current_step.id
      end
      @attachments << att
      self
    end

    def attach_spans(spans)
      @otel_spans = spans
      self
    end

    def get_meta
      StoryMeta.new(
        scenario: @scenario,
        steps: @steps,
        tags: @tags,
        tickets: @tickets,
        covers: @covers,
        meta: @meta,
        suite_path: @suite_path,
        docs: @docs.empty? ? nil : @docs,
        source_order: @source_order,
        otel_spans: @otel_spans
      )
    end

    def record(status:, title: nil, suite_path: nil, source_file: nil, source_line: nil, duration_ms: nil, error: nil)
      @end_time = Process.clock_gettime(Process::CLOCK_MONOTONIC) * 1000.0
      @suite_path = suite_path if suite_path
      duration = duration_ms || (@end_time - @start_time)

      step_events = @steps.each_with_index.map do |s, i|
        s.duration_ms ? RawStepEvent.new(index: i, title: s.text, duration_ms: s.duration_ms) : nil
      end.compact

      tc = RawTestCase.new(
        status: status,
        title: title || @scenario,
        title_path: suite_path ? suite_path + [@scenario] : [@scenario],
        story: get_meta,
        source_file: source_file,
        source_line: source_line,
        duration_ms: duration,
        error: error,
        retry: 0,
        retries: 0,
        attachments: @attachments.empty? ? nil : @attachments,
        step_events: step_events.empty? ? nil : step_events,
        start_time: @start_time,
        end_time: @end_time
      )
      Collector.record(tc)
      tc
    end

    private

    def add_step(keyword, text)
      effective = keyword
      if %w[Given When Then].include?(keyword)
        if @seen_primary[keyword]
          effective = "And"
        else
          @seen_primary[keyword] = true
        end
      end

      step = StoryStep.new(
        id: "step-#{@step_counter}",
        keyword: effective,
        text: text,
        mode: nil,
        wrapped: nil,
        duration_ms: nil,
        docs: nil
      )
      @step_counter += 1
      @steps << step
      @current_step = step
      nil
    end

    def add_explicit_step(keyword, text)
      step = StoryStep.new(
        id: "step-#{@step_counter}",
        keyword: keyword,
        text: text,
        mode: nil,
        wrapped: nil,
        duration_ms: nil,
        docs: nil
      )
      @step_counter += 1
      @steps << step
      @current_step = step
      nil
    end

    def attach_doc(entry)
      if entry["children"] && !entry["children"].empty?
        child_set = entry["children"].to_set(&:object_id)
        filter_docs = ->(docs) { docs.reject { |d| child_set.include?(d.object_id) } }
        @docs = filter_docs.call(@docs)
        @steps.each { |s| s.docs = filter_docs.call(s.docs) if s.docs }
      end

      if @current_step
        @current_step.docs ||= []
        @current_step.docs << entry
      else
        @docs << entry
      end
    end

    def normalize_tickets(ticket)
      return nil unless ticket

      tickets = Array(ticket)
      tickets.map do |t|
        if t.is_a?(String)
          Ticket.new(id: t)
        else
          Ticket.new(id: t[:id] || t["id"], url: t[:url] || t["url"])
        end
      end
    end

    def bridge_otel
      require "opentelemetry-api"
      context = OpenTelemetry::Trace.current_span_context
      return unless context&.valid?

      inject_otel_meta(context)
      inject_otel_docs(context)
      tag_otel_span
    rescue LoadError
      # opentelemetry-api not available
    rescue StandardError
      # OTel not configured, ignore
    end

    def inject_otel_meta(context)
      @meta ||= {}
      @meta["otel"] = { "traceId" => context.trace_id, "spanId" => context.span_id }
    end

    def inject_otel_docs(context)
      @docs << DocEntry.kv("Trace ID", context.trace_id)

      template = @trace_url_template || ENV["OTEL_TRACE_URL_TEMPLATE"]
      return unless template && !template.empty?

      url = template.gsub("{traceId}", context.trace_id)
      @docs << DocEntry.link("View Trace", url)
    end

    def tag_otel_span
      span = OpenTelemetry::Trace.current_span
      return unless span && !span.recording?

      span.set_attribute("story.scenario", @scenario)
      span.set_attribute("story.tags", @tags.join(",")) if @tags && !@tags.empty?
      return unless @tickets

      span.set_attribute("story.tickets", @tickets.map(&:id).join(","))
    end
  end

  module_function

  def init(scenario, **opts)
    Story.new(scenario, **opts)
  end

  # Records a scenario that is specified but not built yet. It appears in the
  # report marked "planned" and stops being planned once someone writes it as a
  # real story with init.
  #
  #   def test_checkout_blocks_suspended_account
  #     ExecutableStories.planned("checkout is blocked for a suspended account")
  #   end
  #
  # Minitest's skip means "do not run this now", which is a different claim from
  # "we have not built this yet", so planned does not skip for you.
  #
  # Recorded at the point of the call: Minitest has no per-test hook to revisit
  # the outcome, so keep this the only statement in the test. Anything after it
  # that fails is recorded by your own `story.record(status: "fail")` call, not
  # by this one.
  #
  # The source location defaults to the caller, so a planned scenario is grouped
  # with the rest of its file instead of landing under an unknown feature. Pass
  # source_file / source_line to override.
  def planned(scenario, source_file: nil, source_line: nil, **opts)
    location = caller_locations(1, 1)&.first
    story = Story.new(scenario, **opts)
    story.record(
      status: "todo",
      duration_ms: 0,
      source_file: source_file || location&.path,
      source_line: source_line || location&.lineno
    )
    story
  end
end
